### PR TITLE
feat: RoiAlign Forward

### PR DESCRIPTION
**Pull request by benjamim93**
## What changes were proposed in this pull request?
Enable RoiAlign forward for MaskRCNN

## How was this patch tested?
Unit tests

## Related links or issues (optional)
Jenkins passed: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/754/

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.utils.{T, Table}
+import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+
+import scala.reflect._
+
+class Pooler[T: ClassTag] (
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Add comments

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    val xhigh = roi.valueAt(4)
+    val yhigh = roi.valueAt(5)
+
+    val area = ev.times(ev.plus(ev.minus(xhigh, xlow), ev.fromType(1)),
+      ev.plus(ev.minus(yhigh, ylow), ev.fromType(1)))
+    area
+  }
+
+  override def updateOutput(input: Table): Tensor[T] = {
+    val feature_maps = input[Table](1)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: featureMaps

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    output.resize(num_rois, num_channels, resolution, resolution)
+      .fill(ev.fromType[Float](Float.MinValue))
+
+    for (level <- 0 until num_levels) {
+      val feature_per_level = feature_maps.get[Tensor[T]](level + 1).get
+      val rois_ind_per_level = roi_levels.zipWithIndex.filter(_._1 == level).map(_._2)
+      val num_rois_per_level = rois_ind_per_level.length
+
+      if (num_rois_per_level > 0) {
+        val rois_per_level = Tensor[T](Array(num_rois_per_level, 5))
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Create new tensor for each time?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
+    gradInput = null
+    gradInput
+  }
+
+  override def toString: String = "nn.Pooler"
+
+  override def clearState(): this.type = {
+    super.clearState()
+    this
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: clear state for each ROIAlign in pooler array

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala (hunk 1 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala (hunk 2 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala (hunk 3 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: ROIs which belong to the current level are extracted here. For each level, the number of ROI could be different. Thus the shape of the Tensor here changes accordingly. I wonder if there is more convenient way to do this.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala (hunk 4 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Added

**Comment by benjamim93** [comment]
Looks like in pyspark/bigdl/nn/layer.py forward() does not support nested lists as input, python api for Pooler is removed and python test is skipped.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        for (i <- 0 until num_rois_per_level) {
+          output(rois_ind_per_level(i) + 1) = res(i + 1)
+        }
+      }
+    }
+
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: throw exception if not support backward

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+        spatialScale)
+    } else {
+      throw new IllegalArgumentException("currently only Double and Float types are supported")
+    }
+
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
+    gradInput = null
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: throw exception

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala (hunk 5 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Added

Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pooler.scala]: Added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala (hunk 6 above)
Comment (benjamim93) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala]: Added

**Pull request merged**